### PR TITLE
MRG+1: fix for set_bipolar_reference

### LINKED ISSUE #5772: [Might be a bug] raw.add_channels
If I have a dateset with only bipolar references, and then I try to append new channels to a raw dataset I get this error:
```
RuntimeError: Don't know how to merge 'custom_ref_applied'. Make sure values are compatible, got types:
    [<class 'bool'>, <class 'bool'>]
```
If I append the channels before making the bipolar references everything is ok, therefore I did not investigate further what was happening in this case :see_no_evil:.

### PR DESCRIPTION
**Pull request by cmmoenne**
I messed up the original pull request while trying to squash the commits (sorry!) :'(

Short description of the bug: `set_bipolar reference` try to remove a list of repeated channels when producing all the bipolar references of a dataset. Also, does not take into account channels of cathode list when performing in-place operations.

In this new pull, I added the requested changes, including a test, and also I updated whats new file.

Note that the current implementation of set_bipolar_reference fail in both, generating all the bipolar combinations and with repeated channels in the anode and cathode lists.

Closes #5780

Closes #5772

### CODE REVIEW

File: doc/whats_new.rst [hunk 1]
```diff
@@ -26,6 +26,10 @@ Changelog
 Bug
 ~~~
 
+- Fix :func:`set_bipolar_reference` in the case of generating all bipolar combinations and also in the case of repeated channels in both lists (anode and cathode) by `Cristóbal Moënne-Loccoz`_
+
+- Fix missing code for computing the median when ``method='median'`` in :meth:`mne.Epochs.average` by `Cristóbal Moënne-Loccoz`_
```
Comment (agramfort) [inline doc/whats_new.rst]: this entry is not related to this PR

File: mne/io/reference.py [hunk 2]
```diff
@@ -522,11 +521,11 @@ def set_bipolar_reference(inst, anode, cathode, ch_name=None, ch_info=None,
...
         inst.info._update_redundant()
-        if an in rem_ca:
-            idx = rem_ca.index(an)
-            del rem_ca[idx]
 
-    # Drop remaining cathode channels
-    inst.drop_channels(rem_ca)
+    # Drop remaining channels.
+    inst.drop_channels([
+        ch_name for ch_name in set(list(anode) + list(cathode))
```
Comment (jona-sassenhagen) [inline mne/io/reference.py]: would the somewhat more concise `for ch_name in set(anode) + set(cathode)` also work?

Comment (jona-sassenhagen) [reply mne/io/reference.py]: ... or, couldn't you just do a set union? `drop_channels = (set(anode) + set(cathode)).union(set(inst.ch_names))` or so

File: mne/io/tests/test_reference.py [hunk 3]
```diff
@@ -473,4 +474,41 @@ def test_add_reference():
...
+    # test classic EOG/ECG bipolar reference (only two channels per pair).
+    raw_test = set_bipolar_reference(raw, ['CH2'], ['CH1'], copy=True)
+    _check_bipolar(raw_test, 'CH2', 'CH1')
+
+    # test all combinations.
+    bipolars = list(itertools.combinations(channels, 2))
+    a_channels, b_channels = [], []
+    for ch_a, ch_b in bipolars:
+        a_channels.append(ch_a)
+        b_channels.append(ch_b)
```
Comment (jona-sassenhagen) [inline mne/io/tests/test_reference.py]: `a_channels, b_channels =*zip(itertools.combinations(channels, 2))`?

File: mne/io/tests/test_reference.py [hunk 4]
```diff
@@ -473,4 +474,41 @@ def test_add_reference():
...
+        ch_names=channels, sfreq=1000., ch_types=['eeg'] * len(channels))
+    raw_data = np.random.randn(len(channels), 1000)
+    raw = RawArray(raw_data, info)
+
+    def _check_bipolar(raw_test, ch_a, ch_b):
+        assert_array_equal(
+            raw_test.get_data(
+                picks=[raw_test.ch_names.index(ch_a + '-' + ch_b)])[0, :],
+            raw_data[channels.index(ch_a), :] -
+            raw_data[channels.index(ch_b), :])
```
Comment (jona-sassenhagen) [inline mne/io/tests/test_reference.py]: I find this a bit hard to read, maybe unpacking it would be a bit nicer.

```
picks = ...
get_data_res = raw.get_data(pick=...)
manual_a = raw_data[channels.index(...
manual_b = ...
assert_array_equal(get_data_res, manual_a - manual_b)
```

or so

File: mne/io/reference.py (hunk 2 above)
Comment (cmmoenne) [reply mne/io/reference.py]: Using only set operations, we could do the following:
```
drop_channels = (set(anode) | set(cathode)) & set(inst.ch_names)
inst.drop_channels(drop_channels)
``` 
I like this option :)

File: mne/io/tests/test_reference.py (hunk 3 above)
Comment (cmmoenne) [reply mne/io/tests/test_reference.py]: The closest expression using zip would be this one (not 100% sure):
```
a_channels, b_channels = [
        list(c) for c in zip(*itertools.combinations(channels, 2))
]
```
...is not as direct as the original, which one would be better?

Comment (jona-sassenhagen) [reply mne/io/tests/test_reference.py]: Ah yes, I mixed up the order of `zip` and the star ...

```
a_chans, b_chans = zip(*itertools.combinations(channels, 2))
raw_test = set_bipolar_reference(raw, list(a_channels), list(b_channels), copy=True)
```

Right?

**Comment by agramfort** [comment]
@jona-sassenhagen we let you do the first round of review if it's ok with you. Tell us when you give a +1. thanls

File: mne/io/tests/test_reference.py (hunk 3 above)
Comment (cmmoenne) [reply mne/io/tests/test_reference.py]: As we need the channels to compare the results after `set_bipolar_reference`, the other alternative is unroll the for (otherwise the generator code should be repeated): 
```
a_channels, b_channels = zip(*itertools.combinations(channels, 2))
a_channels = list(a_channels)
b_channels = list(b_channels)
```
I think this would be the best.

File: mne/io/tests/test_reference.py [hunk 5]
```diff
@@ -483,25 +483,23 @@ def test_bipolar_combinations():
...
     # test all combinations.
-    bipolars = list(itertools.combinations(channels, 2))
-    a_channels, b_channels = [], []
-    for ch_a, ch_b in bipolars:
-        a_channels.append(ch_a)
-        b_channels.append(ch_b)
-    raw_test = set_bipolar_reference(
+    a_channels, b_channels = zip(*itertools.combinations(channels, 2))
+    a_channels = list(a_channels)
+    b_channels = list(b_channels)
```
Comment (jona-sassenhagen) [inline mne/io/tests/test_reference.py]: `a_channels, b_channels = list(a_channels), list(b_channels)`

:)

Ok, that's not necessary. Are you happy with it as is?

Comment (cmmoenne) [reply mne/io/tests/test_reference.py]: I think it's better in a single line :)

**Comment by jona-sassenhagen** [comment]
I've restarted Travis. It was stuck.

File: mne/io/reference.py [hunk 6]
```diff
@@ -522,11 +521,9 @@ def set_bipolar_reference(inst, anode, cathode, ch_name=None, ch_info=None,
...
         inst.info._update_redundant()
-        if an in rem_ca:
-            idx = rem_ca.index(an)
-            del rem_ca[idx]
 
-    # Drop remaining cathode channels
-    inst.drop_channels(rem_ca)
+    # Drop remaining channels.
+    drop_channels = (set(anode) | set(cathode)) & set(inst.ch_names)
+    inst.drop_channels(drop_channels)
```
[the remark this answers is not in the document]
Comment (jona-sassenhagen) [reply mne/io/reference.py]: @cmmoenne I think that would be the place for it, yes.

See https://github.com/mne-tools/mne-python/issues/3183

[the remark this answers is not in the document]
Comment (cmmoenne) [reply mne/io/reference.py]: #3183 error is the same I found in #5772, but seems unrelated to applying the reference and keeping the channels inside `set_bipolar_reference`.

File: mne/io/reference.py [hunk 7]
```diff
@@ -430,6 +430,8 @@ def set_bipolar_reference(inst, anode, cathode, ch_name=None, ch_info=None,
         This parameter can be used to supply a dictionary (or a dictionary for
         each bipolar channel) containing channel information to merge in,
         overwriting the default values. Defaults to None.
+    drop_refs : bool
+        Whether to drop or not the anode/cathode channels from the instance.
```
Comment (mmagnuski) [inline mne/io/reference.py]: nitpick: I think "or not" is unnecessary

File: mne/io/tests/test_reference.py [hunk 8]
```diff
@@ -473,4 +474,48 @@ def test_add_reference():
     pytest.raises(ValueError, add_reference_channels, raw, 1)
 
 
+def test_bipolar_combinations():
+    """Test bipolar channel generation."""
+    channels = ['CH' + str(ni + 1) for ni in range(10)]
```
Comment (mmagnuski) [inline mne/io/tests/test_reference.py]: nitpick: by convention that would be `ch_names`

File: mne/io/tests/test_reference.py [hunk 9]
```diff
@@ -473,4 +474,48 @@ def test_add_reference():
...
+        get_data_res = raw_test.get_data(picks=picks)[0, :]
+        manual_a = raw_data[channels.index(ch_a), :]
+        manual_b = raw_data[channels.index(ch_b), :]
+        assert_array_equal(get_data_res, manual_a - manual_b)
+
+    # test classic EOG/ECG bipolar reference (only two channels per pair).
+    raw_test = set_bipolar_reference(raw, ['CH2'], ['CH1'], copy=True)
+    _check_bipolar(raw_test, 'CH2', 'CH1')
+
+    # test all combinations.
```
Comment (mmagnuski) [inline mne/io/tests/test_reference.py]: is it important to check *all* combinations?

Comment (mmagnuski) [reply mne/io/tests/test_reference.py]: oh, sorry, I read further now, and undrestand that :)

**Comment by mmagnuski** [comment]
I'm almost happy (but the nitpicks you can ignore, it's up to you).

**Comment by jona-sassenhagen** [comment]
I restarted Circle.

**Comment by jona-sassenhagen** [comment]
Will merge if circle is clean this time.

**Comment by mmagnuski** [comment]
Circle is red because of `Invalid requirement: 'dipy --only-binary dipy'` - so rebasing would fix, but I don't think it's necessary when everything else passes.

**Comment by jona-sassenhagen** [comment]
Ah I see. I am merging then. Thanks @mmagnuski

**Comment by jona-sassenhagen** [comment]
... and thanks @cmmoenne ofc :)
